app/src/main/java/org/supla/android/widget/lightswitch/LightSwitchCommandWorker.kt

```diff
@@ -0,0 +1,107 @@
+import org.supla.android.R
+import org.supla.android.SuplaApp
+import org.supla.android.lib.SuplaClient
+import org.supla.android.lib.SuplaConst
+import org.supla.android.widget.SwitchPreferences
+
+const val ARG_TURN_ON = "ARG_TURN_ON"
+
+/**
+ * Worker which is implemented for turning on/off light switches. It supports following channel functions:
+ * light switch [SuplaConst.SUPLA_CHANNELFNC_LIGHTSWITCH],
+ * dimmer [SuplaConst.SUPLA_CHANNELFNC_DIMMER],
+ * RGB lightning [SuplaConst.SUPLA_CHANNELFNC_RGBLIGHTING],
+ * dimmer with RGB lightning [SuplaConst.SUPLA_CHANNELFNC_DIMMERANDRGBLIGHTING]
+ */
+class LightSwitchCommandWorker(appContext: Context, workerParams: WorkerParameters) : Worker(appContext, workerParams) {
+
+    val handler = Handler(Looper.getMainLooper())
+
+    override fun doWork(): Result {
+        val suplaClient = getSuplaClient()
+        if (suplaClient == null) {
+            showToast(R.string.widget_light_switch_command_error)
+            return Result.failure()
+        }
+
+        val widgetIds: IntArray? = inputData.getIntArray(AppWidgetManager.EXTRA_APPWIDGET_IDS)
+        if (widgetIds == null || widgetIds.size != 1) {
+            showToast(R.string.widget_light_switch_command_error)
+            return Result.failure()
+        }
+
+        val configuration = SwitchPreferences(applicationContext).getWidgetConfiguration(widgetIds[0])
+        if (configuration == null) {
+            showToast(R.string.widget_light_switch_command_error)
+            return Result.failure()
+        }
+
+        showToast(R.string.widget_light_switch_command_started)
+
+        if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_LIGHTSWITCH) {
+            suplaClient.turnOnOff(applicationContext, inputData.getBoolean(ARG_TURN_ON, false),
+                    configuration.channelId, false, configuration.channelFunction, false)
+        } else if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_DIMMER
+                || configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_DIMMERANDRGBLIGHTING) {
+            suplaClient.setRGBW(configuration.channelId, configuration.channelColor, 0, getBrightness(), true)
+        } else if (configuration.channelFunction == SuplaConst.SUPLA_CHANNELFNC_RGBLIGHTING) {
+            val brightness = getBrightness()
+            suplaClient.setRGBW(configuration.channelId, configuration.channelColor, brightness, brightness, true)
+        }
+
+        return Result.success()
+    }
+
+    private fun getSuplaClient(): SuplaClient? {
+        val suplaApp = if (applicationContext is SuplaApp) {
+            (applicationContext as SuplaApp)
+        } else {
+            SuplaApp.getApp()
+        }
+
+        if (suplaApp.suplaClient == null) {
+            suplaApp.SuplaClientInitIfNeed(applicationContext)
+        }
+        return suplaApp.suplaClient
+    }
+
+    private fun getBrightness(): Int {
+        return if (inputData.getBoolean(ARG_TURN_ON, false)) {
```

Comment:
It seems to me that the user question at onoff about the brightness level is unnecessary. However, if you want to ask about it, you should also ask about the color.

Reply:
It's not user question. Widget has two buttons -> on and off. For on I'm setting brightness to 100 and for off to 0.